File: algorithms/java/src/test/java/leetcode/LC066PlusOne.java
package leetcode;

import org.junit.Test;

import static org.junit.Assert.assertArrayEquals;

/**
 * https://leetcode.com/problems/plus-one/
 * #easy
 */
public class LC066PlusOne {
    public final class Solution {
        public int[] plusOne(int[] digits) {
            int carry = 1;
            for (int i = digits.length - 1; i >= 0; i--) {
                digits[i] += carry;
                carry = digits[i] / 10;
                digits[i] = digits[i] % 10;
            }
            if (carry > 0) {
                int[] bigger = new int[digits.length + 1];
                bigger[0] = carry;
                System.arraycopy(digits, 0, bigger, 1, digits.length);
                return bigger;
            } else {
                return digits;
            }
        }
    }

    @Test
    public void test_1_9() throws Exception {
        int[] expected = {2, 0};
        assertArrayEquals(expected, new Solution().plusOne(new int[]{1, 9}));
    }

    @Test
    public void test_9_9() throws Exception {
        int[] expected = {1, 0, 0};
        assertArrayEquals(expected, new Solution().plusOne(new int[]{9, 9}));
    }
}

File: algorithms/code/leetcode/lc066_plus_one/lc066_plus_one.py
# https://leetcode.com/problems/plus-one/

import unittest


class Solution:
    def plusOne(self, digits: list[int]) -> list[int]:
        carry = 1
        for i in range(len(digits) - 1, -1, -1):
            digits[i] += carry
            carry = digits[i] // 10
            digits[i] = digits[i] % 10
        if carry > 0:
            bigger = digits.copy()
            bigger.insert(0, carry)
            return bigger
        return digits


class TestCode(unittest.TestCase):
    def test_1_9(self) -> None:
        expected = [2, 0]
        self.assertListEqual(expected, Solution().plusOne([1, 9]))

    def test_9_9(self) -> None:
        expected = [1, 0, 0]
        self.assertListEqual(expected, Solution().plusOne([9, 9]))
